lib/vauth/cleartext.c

```diff
@@ -49,45 +49,50 @@
  * Parameters:
  *
  * data    [in]     - The session handle.
- * userp   [in]     - The user name.
- * passwdp [in]     - The user's password.
+ * authzid [in]     - The authorization identity.
+ * authcid [in]     - The authentication identity.
+ * passwd  [in]     - The password.
  * outptr  [in/out] - The address where a pointer to newly allocated memory
  *                    holding the result will be stored upon completion.
  * outlen  [out]    - The length of the output message.
  *
  * Returns CURLE_OK on success.
  */
 CURLcode Curl_auth_create_plain_message(struct Curl_easy *data,
-                                        const char *userp,
-                                        const char *passwdp,
+                                        const char *authzid,
+                                        const char *authcid,
+                                        const char *passwd,
                                         char **outptr, size_t *outlen)
 {
   CURLcode result;
   char *plainauth;
-  size_t ulen;
+  size_t zlen;
+  size_t clen;
   size_t plen;
   size_t plainlen;
 
   *outlen = 0;
   *outptr = NULL;
-  ulen = strlen(userp);
-  plen = strlen(passwdp);
+  zlen = strlen(authzid);
+  clen = strlen(authcid);
+  plen = strlen(passwd);
 
   /* Compute binary message length. Check for overflows. */
-  if((ulen > SIZE_T_MAX/4) || (plen > (SIZE_T_MAX/2 - 2)))
+  if((zlen > SIZE_T_MAX/2) || (clen > SIZE_T_MAX/2) ||
+     (plen > (SIZE_T_MAX/2 - 2)))
```

Comment:
The logic here looks wrong. The total can not be allowed to be larger than `SIZE_T_MAX` when we sum them all up and add 2.

Reply:
Yeah I didn't like that logic to begin with and I think I have made it worse with potentially beginning back the CVE :(

Dividing the addition of zlen and clen by 4 would give us the same functionality as before but...

However the RFC does state that servers must be capable of accepting up to and including 255 octets. I appreciate this particular code has changed a little over time (945f60e8, c136657 and f3a24d79) but is there a reason why we use SIZE_T_MAX as opposed to 255? Admittedly the RFC doesn't state whether larger than 255 is valid.

Reply:
It's just that we don't normally artificially limit these things so we haven't done so here either. The check is only there to make sure we don't get a integer overflow in the math after it.

I wouldn't mind setting a limit somewhere to make the check more straight forward. The question is then only what a sensible limit should be? 255 seems rather short maximum with today's standards. 1024? 10240? 102400? I don't think we have any good reason to have a very low limit...

Reply:
Good question. I will have a think and see if I can find anything out for some of the free/opensource/freeware mail servers out there.

In the meantime I've updated my patch as per my comment above.